### PR TITLE
CMSIS-NN: Add elementwise add/mul int16 and unit tests

### PR DESCRIPTION
**Pull request by mansnils**
Co-authored-by: Annie Tallund <annie.tallund@arm.com> and Måns Nilsson <mans.nilsson@arm.com>

### CODE REVIEW

**Comment by grasci-arm** [comment]
Can one of the admins verify this patch?

File: CMSIS/NN/Include/arm_nnfunctions.h [hunk 1]
```diff
@@ -1684,27 +1684,27 @@ extern "C" {
...
+ * @param[in]       input_1_offset          offset for input 1. Range: -127 to 128
  * @param[in]       input_1_mult            multiplier for input 1
  * @param[in]       input_1_shift           shift for input 1
- * @param[in]       input_2_offset          offset for input 2. Range: Range: -127 to 128
+ * @param[in]       input_2_offset          offset for input 2. Range: -127 to 128
  * @param[in]       input_2_mult            multiplier for input 2
  * @param[in]       input_2_shift           shift for input 2
  * @param[in]       left_shift              input left shift
  * @param[in,out]   output                  pointer to output vector
  * @param[in]       out_offset              output offset
```
Comment (felix-johnny) [inline CMSIS/NN/Include/arm_nnfunctions.h:1622]: Please add Range for output offset as -128 to 127

File: CMSIS/NN/Include/arm_nnfunctions.h [hunk 2]
```diff
@@ -1684,27 +1684,27 @@ extern "C" {
...
  * @param[in]       out_mult                output multiplier
  * @param[in]       out_shift               output shift
- * @param[in]       out_activation_min      minimum value to clamp output to
- * @param[in]       out_activation_max      maximum value to clamp output to
+ * @param[in]       out_activation_min      minimum value to clamp output to. Min: -127
```
Comment (felix-johnny) [inline CMSIS/NN/Include/arm_nnfunctions.h:1625]: This should be -128 to 127. It is only the input offset that is passed on as -input_offset making the range as -127 to 128.

File: CMSIS/NN/Include/arm_nnfunctions.h [hunk 3]
```diff
@@ -1723,20 +1723,57 @@ arm_status arm_elementwise_add_s8(const int8_t *input_1_vect,
...
+                                  const int32_t block_size);
 
 /**
- * @brief s8 element wise multiplication
+ * @brief s16 elementwise add of two vectors
  * @param[in]       input_1_vect            pointer to input vector 1
  * @param[in]       input_2_vect            pointer to input vector 2
- * @param[in]       input_1_offset          offset for input 1. Range: Range: -127 to 128
- * @param[in]       input_2_offset          offset for input 2. Range: Range: -127 to 128
+ * @param[in]       input_1_offset          offset for input 1. Range:
```
Comment (felix-johnny) [inline CMSIS/NN/Include/arm_nnfunctions.h:1651]: Please fill up range here. The same logic as for int8 applies here. -input_offset is passed on.

File: CMSIS/NN/Include/arm_nnfunctions.h [hunk 4]
```diff
@@ -1723,20 +1723,57 @@ arm_status arm_elementwise_add_s8(const int8_t *input_1_vect,
...
+ * @param[in]       input_1_mult            multiplier for input 1
+ * @param[in]       input_1_shift           shift for input 1
+ * @param[in]       input_2_offset          offset for input 2. Range:
```
Comment (felix-johnny) [inline CMSIS/NN/Include/arm_nnfunctions.h:1654]: Please add range here.

File: CMSIS/NN/Include/arm_nnfunctions.h [hunk 5]
```diff
@@ -1723,20 +1723,57 @@ arm_status arm_elementwise_add_s8(const int8_t *input_1_vect,
...
+ * @param[in]       input_2_vect            pointer to input vector 2
+ * @param[in]       input_1_offset          offset for input 1. Range: -127 to 128
+ * @param[in]       input_2_offset          offset for input 2. Range: -127 to 128
  * @param[in,out]   output                  pointer to output vector
  * @param[in]       out_offset              output offset
  * @param[in]       out_mult                output multiplier
  * @param[in]       out_shift               output shift
- * @param[in]       out_activation_min      minimum value to clamp output to
- * @param[in]       out_activation_max      maximum value to clamp output to
+ * @param[in]       out_activation_min      minimum value to clamp output to. Min: -127
```
Comment (felix-johnny) [inline CMSIS/NN/Include/arm_nnfunctions.h:1694]: same comment as for add.

File: CMSIS/NN/Include/arm_nnfunctions.h [hunk 6]
```diff
@@ -1752,7 +1789,37 @@ arm_status arm_elementwise_mul_s8(const int8_t *input_1_vect,
...
                                   const int32_t out_activation_min,
                                   const int32_t out_activation_max,
-                                  const uint32_t block_size);
+                                  const int32_t block_size);
+
+/**
+ * @brief s16 elementwise multiplication
+ * @param[in]       input_1_vect            pointer to input vector 1
+ * @param[in]       input_2_vect            pointer to input vector 2
+ * @param[in]       input_1_offset          offset for input 1. Range: -32768 to 32767
```
Comment (felix-johnny) [inline CMSIS/NN/Include/arm_nnfunctions.h:1717]: -input_offset is passed on to the API.

File: CMSIS/NN/Include/arm_nnfunctions.h [hunk 7]
```diff
@@ -1752,7 +1789,37 @@ arm_status arm_elementwise_mul_s8(const int8_t *input_1_vect,
...
+ * @param[in]       input_2_offset          offset for input 2. Range: -32768 to 32767
+ * @param[in,out]   output                  pointer to output vector
+ * @param[in]       out_offset              output offset
```
Comment (felix-johnny) [inline CMSIS/NN/Include/arm_nnfunctions.h:1720]: Could you double check if output offset is valid for int16?

File: CMSIS/NN/Include/arm_nnfunctions.h (hunk 1 above)
Comment (mansnils) [reply CMSIS/NN/Include/arm_nnfunctions.h]: Right

File: CMSIS/NN/Include/arm_nnfunctions.h (hunk 2 above)
Comment (mansnils) [reply CMSIS/NN/Include/arm_nnfunctions.h]: Right

File: CMSIS/NN/Include/arm_nnfunctions.h (hunk 3 above)
Comment (mansnils) [reply CMSIS/NN/Include/arm_nnfunctions.h]: Done

File: CMSIS/NN/Include/arm_nnfunctions.h (hunk 4 above)
Comment (mansnils) [reply CMSIS/NN/Include/arm_nnfunctions.h]: Done

File: CMSIS/NN/Include/arm_nnfunctions.h (hunk 5 above)
Comment (mansnils) [reply CMSIS/NN/Include/arm_nnfunctions.h]: Done

File: CMSIS/NN/Include/arm_nnfunctions.h (hunk 6 above)
Comment (mansnils) [reply CMSIS/NN/Include/arm_nnfunctions.h]: Done

**Review by felix-johnny — APPROVED** [review]

**Pull request merged**